File: tests/netDxf.Conformance/HatchScalarOrderTests.cs
using netDxf;
using netDxf.Entities;
using netDxf.Header;
using netDxf.IO;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static void RegisterHatchScalarOrderTests()
    {
        foreach (DxfVersion version in SupportedVersions)
            foreach (bool binary in new[] { false, true })
                foreach (short kind in new short[] { 1, 2, 3 })
                {
                    DxfVersion v = version; bool b = binary; short k = kind;
                    Run($"hatch/scalar-order/permutations/{v}/{b}/{k}", () => HatchScalarPermutations(v, b, k));
                    Run($"hatch/scalar-order/missing-duplicate/{v}/{b}/{k}", () => HatchScalarInvalid(v, b, k));
                    Run($"hatch/scalar-order/next-edge/{v}/{b}/{k}", () => HatchScalarNextEdge(v, b, k));
                }
        foreach (DxfVersion version in SupportedVersions)
            foreach (bool binary in new[] { false, true })
                foreach (bool spline in new[] { false, true })
                {
                    DxfVersion v = version; bool b = binary, s = spline;
                    Run($"hatch/scalar-order/header/{v}/{b}/{s}", () => HatchUnorderedHeader(v, b, s));
                }
    }

    private static List<DxfTag> HatchScalarTags(DxfVersion version, short kind, int shift, bool reverse)
    {
        var tags = HatchEdgeDispatchTags(version, kind);
        int start = tags.FindIndex(t => t.Code == 72) + 1, end = tags.FindIndex(t => t.Code == 97);
        var packet = tags.GetRange(start, end - start);
        if (kind == 1)
        {
            packet[0] = new(10, -0.125); packet[1] = new(20, 2.0000000000000004);
        }
        if (reverse) packet.Reverse();
        var reordered = packet.Skip(shift).Concat(packet.Take(shift)).ToList();
        tags.RemoveRange(start, end - start); tags.InsertRange(start, reordered);
        return tags;
    }

    private static void AssertHatchScalar(HatchBoundaryPath.Edge edge, short kind)
    {
        Equal((HatchBoundaryPath.EdgeType)kind, edge.Type, "Scalar edge kind");
        switch (edge)
        {
            case HatchBoundaryPath.Line line:
                SameDoubleBits(-0.125, line.Start.X, "Line start X"); SameDoubleBits(2.0000000000000004, line.Start.Y, "Line start Y");
                Equal(new Vector2(10, 10), line.End, "Line end"); break;
            case HatchBoundaryPath.Arc arc:
                Equal(new Vector2(1, 2), arc.Center, "Arc center"); Equal(3.0, arc.Radius, "Arc radius");
                Equal(0.0, arc.StartAngle, "Arc start"); Equal(360.0, arc.EndAngle, "Arc end"); Check(arc.IsCounterclockwise, "Arc direction"); break;
            case HatchBoundaryPath.Ellipse ellipse:
                Equal(new Vector2(1, 2), ellipse.Center, "Ellipse center"); Equal(new Vector2(3, 0), ellipse.EndMajorAxis, "Ellipse major vector");
                Equal(0.5, ellipse.MinorRatio, "Ellipse ratio"); Equal(0.0, ellipse.StartAngle, "Ellipse start");
                Equal(360.0, ellipse.EndAngle, "Ellipse end"); Check(ellipse.IsCounterclockwise, "Ellipse direction"); break;
            default: throw new InvalidOperationException("Unexpected scalar edge.");
        }
    }

    private static void HatchScalarPermutations(DxfVersion version, bool binary, short kind)
    {
        int count = kind == 1 ? 4 : kind == 2 ? 6 : 8;
        foreach (bool reverse in new[] { false, true })
            for (int shift = 0; shift < count; shift++)
            {
                var tags = HatchScalarTags(version, kind, shift, reverse);
                if (!binary && reverse)
                {
                    int start = tags.FindIndex(t => t.Code == 72), end = tags.FindIndex(t => t.Code == 97);
                    for (int i = end; i > start; i--) tags.Insert(i, new(999, "72 97 0 ENDSEC"));
                }
                using var input = new MemoryStream(RawFixtureBytes(tags, binary));
                var doc = DxfDocument.Load(input) ?? throw new InvalidOperationException("Reordered HATCH scalar input rejected.");
                for (int cycle = 0; cycle < 2; cycle++)
                {
                    Hatch hatch = doc.Entities.Hatches.Single();
                    AssertHatchScalar(hatch.BoundaryPaths.Single().Edges.Single(), kind);
                    AssertHatchScalar(((Hatch)hatch.Clone()).BoundaryPaths.Single().Edges.Single(), kind);
                    Equal("after pattern", (string)hatch.XData["DOUBLE_TEST"].XDataRecord.Single().Value, "Following XData");
                    Equal(new Vector2(2, 3), hatch.SeedPoints.Single(), "Seed");
                    using var output = new MemoryStream(); Check(doc.Save(output, !binary), "Scalar edge save failed.");
                    if (cycle == 0 && shift == 0 && reverse)
                        File.WriteAllBytes(Path.Combine(ArtifactDirectory, $"hatch-scalar-order-{version}-{binary}-{kind}.dxf"), output.ToArray());
                    output.Position = 0; doc = DxfDocument.Load(output) ?? throw new InvalidOperationException("Scalar edge reload failed.");
                }
                Check(input.CanRead, "Scalar reader closed input.");
            }
    }

    private static void HatchScalarInvalid(DxfVersion version, bool binary, short kind)
    {
        var original = HatchScalarTags(version, kind, 1, true);
        int start = original.FindIndex(t => t.Code == 72) + 1, end = original.FindIndex(t => t.Code == 97);
        for (int i = start; i < end; i++)
        {
            var missing = original.ToList(); missing.RemoveAt(i); ExpectEdgePacketInvalid(missing, binary);
            var duplicate = original.ToList(); duplicate.Insert(end, original[i]); ExpectEdgePacketInvalid(duplicate, binary);
        }
        var unframed = original.ToList(); unframed.Insert(end, new(75, (short)0)); ExpectEdgePacketInvalid(unframed, binary);
        if (kind != 1)
        {
            var flag = original.ToList(); flag[flag.FindIndex(start, t => t.Code == 73)] = new(73, (short)2);
            ExpectEdgePacketInvalid(flag, binary);
        }
        var overrun = original.ToList(); overrun.RemoveAt(end - 1);
        // A following entity cannot satisfy the missing field even when its group codes match.
        overrun.InsertRange(end - 1, new DxfTag[] { new(0, "LINE"), new(5, "ABCD"), new(100, "AcDbEntity"), new(8, "0"), new(100, "AcDbLine") });
        ExpectEdgePacketInvalid(overrun, binary);
    }

    private static void HatchScalarNextEdge(DxfVersion version, bool binary, short kind)
    {
        var tags = HatchScalarTags(version, kind, 1, true);
        int refs = tags.FindIndex(t => t.Code == 97);
        tags[tags.FindIndex(t => t.Code == 93)] = new(93, 2);
        tags.InsertRange(refs, new DxfTag[] { new(72, (short)1), new(21, -7.0), new(11, 8.0), new(20, 5.0), new(10, -6.0) });
        using var input = new MemoryStream(RawFixtureBytes(tags, binary));
        var doc = DxfDocument.Load(input) ?? throw new InvalidOperationException("Two reordered edges rejected.");
        var edges = doc.Entities.Hatches.Single().BoundaryPaths.Single().Edges;
        Equal(2, edges.Count, "Counted edges changed"); AssertHatchScalar(edges[0], kind);
        var line = (HatchBoundaryPath.Line)edges[1];
        Equal(new Vector2(-6, 5), line.Start, "Second edge start"); Equal(new Vector2(8, -7), line.End, "Second edge end");
    }
    private static void HatchUnorderedHeader(DxfVersion version, bool binary, bool spline)
    {
        var source = spline ? HatchEdgeDispatchTags(version, 4) : HatchDoubleTags(version, HatchType.UserDefined, 0);
        int start = spline ? source.FindIndex(t => t.Code == 94) : source.FindIndex(t => t.Code == 92) + 1;
        int count = spline ? 5 : 3;
        var header = source.GetRange(start, count); header.Reverse();
        for (int shift = 0; shift < count; shift++)
        {
            var tags = source.ToList(); tags.RemoveRange(start, count);
            tags.InsertRange(start, header.Skip(shift).Concat(header.Take(shift)));
            if (!binary) tags.Insert(start + 1, new(999, "95 96 93 73"));
            using var input = new MemoryStream(RawFixtureBytes(tags, binary));
            var doc = DxfDocument.Load(input) ?? throw new InvalidOperationException("Reordered list header rejected.");
            for (int cycle = 0; cycle < 2; cycle++)
            {
                var hatch = doc.Entities.Hatches.Single(); var edge = hatch.BoundaryPaths.Single().Edges.Single();
                if (spline)
                {
                    var e = (HatchBoundaryPath.Spline)edge;
                    Equal((short)2, e.Degree, "Reordered degree"); Check(!e.IsRational && !e.IsPeriodic, "Reordered spline flags");
                    Check(e.Knots.SequenceEqual(new double[] { 0, 0, 0, 1, 1, 1 }), "Knot list changed");
                    Equal(new Vector3(5, 10, 1), e.ControlPoints[1], "Control list changed");
                }
                else
                {
                    var e = (HatchBoundaryPath.Polyline)edge;
                    Check(e.IsClosed && e.Vertexes.Length == 4, "Polyline header changed");
                    Equal(new Vector3(10, 10, 0), e.Vertexes[2], "Polyline list changed");
                }
                Equal(new Vector2(2, 3), hatch.SeedPoints.Single(), "Following header seed");
                using var output = new MemoryStream(); Check(doc.Save(output, !binary), "Reordered header save failed");
                output.Position = 0; doc = DxfDocument.Load(output) ?? throw new InvalidOperationException("Reordered header reload failed");
            }
        }
        foreach (int i in Enumerable.Range(0, count))
            foreach (bool duplicate in new[] { false, true })
            {
                var tags = source.ToList();
                if (duplicate) tags.Insert(start + count, source[start + i]); else tags.RemoveAt(start + i);
                using var input = new MemoryStream(RawFixtureBytes(tags, binary));
#if DEBUG
                Throws<InvalidDataException>(() => DxfDocument.Load(input));
#else
                Check(DxfDocument.Load(input) == null, "Malformed list header accepted.");
#endif
            }
    }

}

File: javascript/tests/netDxf.Conformance/HatchScalarOrderTests.js
// Port of the pinned C# conformance module; original case identities/assertions retained.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { Run, Check, Equal, SameDoubleBits, Throws, SupportedVersions, VersionName, BooleanName } from './TestHarness.js';
import { DxfDocument, DxfTag, HatchType, HatchBoundaryPath, MemoryStream, Vector2, Vector3 } from '../../index.js';
import { InvalidDataException } from '../../runtime/Errors.js';
import { GetTypedIOConfiguration } from '../../runtime/TypedDocumentIO.js';
import { HatchDoubleTags } from './HatchDoublePatternTests.js';
import { HatchEdgeDispatchTags, ExpectEdgePacketInvalid } from './HatchEdgeDispatchTests.js';
import { RawFixtureBytes } from './RawDocumentTests.js';
const single = items => { const a = Array.from(items); Equal(1, a.length, 'Expected one item'); return a[0]; };
const artifacts = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../artifacts/conformance/fixtures');
export function RegisterHatchScalarOrderTests() {
  for (const v of SupportedVersions) for (const b of [false, true]) for (const k of [1, 2, 3]) {
    const suffix = `${VersionName(v)}/${BooleanName(b)}/${k}`;
    Run(`hatch/scalar-order/permutations/${suffix}`, () => HatchScalarPermutations(v, b, k));
    Run(`hatch/scalar-order/missing-duplicate/${suffix}`, () => HatchScalarInvalid(v, b, k));
    Run(`hatch/scalar-order/next-edge/${suffix}`, () => HatchScalarNextEdge(v, b, k));
  }
  for (const v of SupportedVersions) for (const b of [false, true]) for (const s of [false, true])
    Run(`hatch/scalar-order/header/${VersionName(v)}/${BooleanName(b)}/${BooleanName(s)}`, () => HatchUnorderedHeader(v, b, s));
}
export function HatchScalarTags(version, kind, shift, reverse) {
  const tags = HatchEdgeDispatchTags(version, kind), start = tags.findIndex(t => t.Code === 72) + 1, end = tags.findIndex(t => t.Code === 97), packet = tags.slice(start, end);
  if (kind === 1) { packet[0] = new DxfTag(10, -.125); packet[1] = new DxfTag(20, 2.0000000000000004); }
  if (reverse) packet.reverse();
  tags.splice(start, end - start, ...packet.slice(shift), ...packet.slice(0, shift)); return tags;
}
export function AssertHatchScalar(edge, kind) {
  Equal(kind, edge.Type, 'Scalar edge kind');
  if (edge instanceof HatchBoundaryPath.Line) {
    SameDoubleBits(-.125, edge.Start.X, 'Line start X'); SameDoubleBits(2.0000000000000004, edge.Start.Y, 'Line start Y'); Equal(new Vector2(10, 10), edge.End, 'Line end');
  } else if (edge instanceof HatchBoundaryPath.Arc) {
    Equal(new Vector2(1, 2), edge.Center, 'Arc center'); Equal(3, edge.Radius, 'Arc radius'); Equal(0, edge.StartAngle, 'Arc start'); Equal(360, edge.EndAngle, 'Arc end'); Check(edge.IsCounterclockwise, 'Arc direction');
  } else if (edge instanceof HatchBoundaryPath.Ellipse) {
    Equal(new Vector2(1, 2), edge.Center, 'Ellipse center'); Equal(new Vector2(3, 0), edge.EndMajorAxis, 'Ellipse major vector'); Equal(.5, edge.MinorRatio, 'Ellipse ratio'); Equal(0, edge.StartAngle, 'Ellipse start'); Equal(360, edge.EndAngle, 'Ellipse end'); Check(edge.IsCounterclockwise, 'Ellipse direction');
  } else throw new Error('Unexpected scalar edge.');
}
export function HatchScalarPermutations(version, binary, kind) {
  const count = kind === 1 ? 4 : kind === 2 ? 6 : 8;
  for (const reverse of [false, true]) for (let shift = 0; shift < count; shift++) {
    const tags = HatchScalarTags(version, kind, shift, reverse);
    if (!binary && reverse) { const start = tags.findIndex(t => t.Code === 72), end = tags.findIndex(t => t.Code === 97); for (let i = end; i > start; i--) tags.splice(i, 0, new DxfTag(999, '72 97 0 ENDSEC')); }
    const input = new MemoryStream(RawFixtureBytes(tags, binary));
    try {
      let doc = DxfDocument.Load(input); Check(doc !== null, 'Reordered HATCH scalar input rejected.');
      for (let cycle = 0; cycle < 2; cycle++) {
        const hatch = single(doc.Entities.Hatches);
        AssertHatchScalar(single(single(hatch.BoundaryPaths).Edges), kind); AssertHatchScalar(single(single(hatch.Clone().BoundaryPaths).Edges), kind);
        Equal('after pattern', single(hatch.XData.get_Item('DOUBLE_TEST').XDataRecord).Value, 'Following XData'); Equal(new Vector2(2, 3), single(hatch.SeedPoints), 'Seed');
        const output = new MemoryStream(); try {
          Check(doc.Save(output, !binary), 'Scalar edge save failed.');
          if (cycle === 0 && shift === 0 && reverse) { fs.mkdirSync(artifacts, { recursive: true }); fs.writeFileSync(path.join(artifacts, `hatch-scalar-order-${VersionName(version)}-${BooleanName(binary)}-${kind}.dxf`), output.ToArray()); }
          output.Position = 0; doc = DxfDocument.Load(output); Check(doc !== null, 'Scalar edge reload failed.');
        } finally { output.Dispose(); }
      }
      Check(input.CanRead, 'Scalar reader closed input.');
    } finally { input.Dispose(); }
  }
}
export function HatchScalarInvalid(version, binary, kind) {
  const original = HatchScalarTags(version, kind, 1, true), start = original.findIndex(t => t.Code === 72) + 1, end = original.findIndex(t => t.Code === 97);
  for (let i = start; i < end; i++) {
    const missing = original.slice(); missing.splice(i, 1); ExpectEdgePacketInvalid(missing, binary);
    const duplicate = original.slice(); duplicate.splice(end, 0, original[i]); ExpectEdgePacketInvalid(duplicate, binary);
  }
  const unframed = original.slice(); unframed.splice(end, 0, new DxfTag(75, 0)); ExpectEdgePacketInvalid(unframed, binary);
  if (kind !== 1) { const flag = original.slice(); flag[flag.findIndex((t, i) => i >= start && t.Code === 73)] = new DxfTag(73, 2); ExpectEdgePacketInvalid(flag, binary); }
  const overrun = original.slice(); overrun.splice(end - 1, 1); overrun.splice(end - 1, 0, ...[[0, 'LINE'], [5, 'ABCD'], [100, 'AcDbEntity'], [8, '0'], [100, 'AcDbLine']].map(([c, v]) => new DxfTag(c, v))); ExpectEdgePacketInvalid(overrun, binary);
}
export function HatchScalarNextEdge(version, binary, kind) {
  const tags = HatchScalarTags(version, kind, 1, true), refs = tags.findIndex(t => t.Code === 97);
  tags[tags.findIndex(t => t.Code === 93)] = new DxfTag(93, 2);
  tags.splice(refs, 0, ...[[72, 1], [21, -7], [11, 8], [20, 5], [10, -6]].map(([c, v]) => new DxfTag(c, v)));
  const input = new MemoryStream(RawFixtureBytes(tags, binary)); try {
    const doc = DxfDocument.Load(input); Check(doc !== null, 'Two reordered edges rejected.');
    const edges = Array.from(single(single(doc.Entities.Hatches).BoundaryPaths).Edges); Equal(2, edges.length, 'Counted edges changed'); AssertHatchScalar(edges[0], kind);
    Equal(new Vector2(-6, 5), edges[1].Start, 'Second edge start'); Equal(new Vector2(8, -7), edges[1].End, 'Second edge end');
  } finally { input.Dispose(); }
}
export function HatchUnorderedHeader(version, binary, spline) {
  const source = spline ? HatchEdgeDispatchTags(version, 4) : HatchDoubleTags(version, HatchType.UserDefined, 0);
  const start = spline ? source.findIndex(t => t.Code === 94) : source.findIndex(t => t.Code === 92) + 1, count = spline ? 5 : 3, header = source.slice(start, start + count).reverse();
  for (let shift = 0; shift < count; shift++) {
    const tags = source.slice(); tags.splice(start, count, ...header.slice(shift), ...header.slice(0, shift)); if (!binary) tags.splice(start + 1, 0, new DxfTag(999, '95 96 93 73'));
    const input = new MemoryStream(RawFixtureBytes(tags, binary)); try {
      let doc = DxfDocument.Load(input); Check(doc !== null, 'Reordered list header rejected.');
      for (let cycle = 0; cycle < 2; cycle++) {
        const hatch = single(doc.Entities.Hatches), edge = single(single(hatch.BoundaryPaths).Edges);
        if (spline) { Equal(2, edge.Degree, 'Reordered degree'); Check(!edge.IsRational && !edge.IsPeriodic, 'Reordered spline flags'); Equal([0, 0, 0, 1, 1, 1], edge.Knots, 'Knot list changed'); Equal(new Vector3(5, 10, 1), edge.ControlPoints[1], 'Control list changed'); }
        else { Check(edge.IsClosed && edge.Vertexes.length === 4, 'Polyline header changed'); Equal(new Vector3(10, 10, 0), edge.Vertexes[2], 'Polyline list changed'); }
        Equal(new Vector2(2, 3), single(hatch.SeedPoints), 'Following header seed');
        const output = new MemoryStream(); try { Check(doc.Save(output, !binary), 'Reordered header save failed'); output.Position = 0; doc = DxfDocument.Load(output); Check(doc !== null, 'Reordered header reload failed'); } finally { output.Dispose(); }
      }
    } finally { input.Dispose(); }
  }
  for (let i = 0; i < count; i++) for (const duplicate of [false, true]) {
    const tags = source.slice(); if (duplicate) tags.splice(start + count, 0, source[start + i]); else tags.splice(start + i, 1);
    const input = new MemoryStream(RawFixtureBytes(tags, binary)); try { if (GetTypedIOConfiguration() === 'Debug') Throws(InvalidDataException, () => DxfDocument.Load(input)); else Check(DxfDocument.Load(input) === null, 'Malformed list header accepted.'); } finally { input.Dispose(); }
  }
}
